File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        dq = deque([(root, 1)])
        
        max_width = 1    
        while (len(dq)):
            n = len(dq)
            curr_width = dq[-1][1] - dq[0][1] + 1
            max_width = max(curr_width, max_width)
            for i in range(n):
                top = dq[0][0]
                idx = dq[0][1]
                
                dq.popleft()
                if (top.left):
                    dq.append((top.left, idx * 2))
                if (top.right):
                    dq.append((top.right, idx * 2 + 1))
        return max_width

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) 
    {
        deque<pair<TreeNode*,unsigned long long int>> dq;
        dq.push_back({root,0});
        int max_width = 1;
        while (!dq.empty())
        {
            int n = dq.size();
            int curr_width = int(dq.back().second - dq.front().second + 1);
            max_width = max(max_width, curr_width);
            while (n--)
            {
                TreeNode *top = dq.front().first;
                unsigned long long int index = dq.front().second;
                dq.pop_front();
                if (top->left)
                    dq.push_back({top->left, 2 * index + 1});
                if (top->right)
                    dq.push_back({top->right, 2 * index + 2});            
            }
        }
        return max_width;
    }
};
